File: core/typescript/src/transpose.ts
/**
 * transpose() — the keystone API of `@cv-transpose/core`.
 *
 * Given a list of input CVs and a tenant template, run the full per-CV
 * pipeline:
 *   1. Extract raw text from the upload (PDF / DOCX / DOC).
 *   2. Call the injected {@link LlmProvider} to produce a structured CvData.
 *   3. Render a DOCX using the manifest-derived template contract.
 *   4. Validate page 1 (overflow / underflow heuristics via LibreOffice).
 *
 * This function is intentionally side-effect free with respect to sessions,
 * SSE, encryption, retries, and quality reviews — those are concerns of the
 * api orchestrator (which will delegate to transpose() for the per-CV path).
 *
 * Errors are captured per-input: a failure on one CV does not abort the
 * batch. Each {@link TransposedCv} carries its own `errors` array (populated
 * on fail-loud paths) and a `warnings` array via `alignmentReport`.
 */

import { readFile, mkdtemp, rm, writeFile } from 'node:fs/promises';
import { tmpdir } from 'node:os';
import { join } from 'node:path';

import { extractTextFromBuffer } from './extract/text.js';
import { validatePage1 } from './validate/page1.js';
import {
  buildTemplateDocumentXml,
  getXmlHeader,
  writeTemplateHeader,
} from './template/render.js';
import { unpackDocx, packDocx } from './docx/tools.js';
import { manifestToContract } from './template/bridge.js';
import {
  deriveOutputNameFromTemplateContract,
  getPrimaryExperienceSectionLabel,
  getPrimarySectorSectionLabel,
} from './template/contract.js';
import { cvDataSchema, type CvData } from './cv/profile.js';
import { defaultLogger as logger } from './util/log.js';
import { buildSystemPrompt, buildUserPrompt } from './transpose-prompt.js';
import type { LlmCompleteArgs } from './llm.js';
import type {
  AlignmentReport,
  DetectedFields,
  TransposeInput,
  TransposeOutput,
  TransposedCv,
} from './types.js';
import type { TemplateContract } from './template/contract.js';

/**
 * Build a streaming-delta forwarder that routes the provider's deltas to the
 * file-scoped streamCallbacks. Created per file so the file name is captured
 * in the closure.
 */
function makeDeltaForwarder(
  fileName: string,
  cb: NonNullable<TransposeInput['streamCallbacks']>,
): NonNullable<LlmCompleteArgs['onDelta']> {
  return (delta) => {
    if (delta.kind === 'thinking') {
      cb.onThinkingDelta?.(fileName, delta.text);
    } else {
      cb.onContentDelta?.(fileName, delta.text);
    }
  };
}

/**
 * Escape XML special characters and a handful of common diacritics that the
 * downstream renderer otherwise emits as raw bytes (which LibreOffice and
 * Word render correctly, but some downstream tools choke on). Mirrors the
 * api orchestrator's `escapeXml` to keep parity until the renderer learns to
 * escape its own inputs.
 */
function escapeXml(text: string): string {
  return text
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/–/g, '&#x2013;') // en-dash
    .replace(/‘/g, '&#x2018;')
    .replace(/’/g, '&#x2019;')
    .replace(/“/g, '&#x201C;')
    .replace(/”/g, '&#x201D;');
}

/**
 * Run a callback inside a fresh mkdtemp dir and clean up on the way out,
 * mirroring the saturation-safe pattern used by `extractTextFromBuffer` and
 * `validatePage1`.
 */
async function withTempDir<T>(prefix: string, fn: (dir: string) => Promise<T>): Promise<T> {
  const dir = await mkdtemp(join(tmpdir(), prefix));
  try {
    return await fn(dir);
  } finally {
    await rm(dir, { recursive: true, force: true });
  }
}

function escapeCvData(data: CvData): CvData {
  return {
    ...data,
    name: escapeXml(data.name),
    title_line1: escapeXml(data.title_line1),
    title_line2: escapeXml(data.title_line2),
    technicalSkills: data.technicalSkills.map((skill) => ({
      label: escapeXml(skill.label),
      description: escapeXml(skill.description),
    })),
    sectors: data.sectors.map((sector) => escapeXml(sector)),
    domains: data.domains.map((domain) => escapeXml(domain)),
    experience: data.experience.map((job) => ({
      ...job,
      company: escapeXml(job.company),
      description: escapeXml(job.description),
      dates: escapeXml(job.dates),
      title: escapeXml(job.title),
      tasks: job.tasks.map((task) => escapeXml(task)),
      achievements: job.achievements.map((achievement) => escapeXml(achievement)),
      techEnvironment: escapeXml(job.techEnvironment),
    })),
    languages: data.languages.map((language) => ({
      label: escapeXml(language.label),
      level: escapeXml(language.level),
    })),
    education: data.education.map((entry) => ({
      year: escapeXml(entry.year),
      description: escapeXml(entry.description),
    })),
  };
}

async function renderDocxFromContract(params: {
  baseDocx: Uint8Array;
  data: CvData;
  contract: TemplateContract;
}): Promise<Uint8Array> {
  return withTempDir('cv-transpose-render-', async (workDir) => {
    // unpackDocx / packDocx are filesystem-based today (Python parity), so
    // we materialise the base.docx to a temp file, unpack into a sibling
    // dir, and pack to another temp file. This mirrors what the api
    // orchestrator does today.
    const baseDocxPath = join(workDir, 'base.docx');
    await writeFile(baseDocxPath, Buffer.from(params.baseDocx));

    const unpackDir = join(workDir, 'unpacked');
    const [unpackOk, unpackMsg] = await unpackDocx(baseDocxPath, unpackDir);
    if (!unpackOk) {
      throw new Error(`unpackDocx failed: ${unpackMsg}`);
    }

    const documentXmlPath = join(unpackDir, 'word', 'document.xml');
    const xmlHeader = await getXmlHeader(documentXmlPath);
    const escaped = escapeCvData(params.data);
    const documentXml = buildTemplateDocumentXml(escaped, params.contract, xmlHeader);
    await writeFile(documentXmlPath, documentXml, 'utf-8');

    // Some DOCX templates (Word 2016+) put the title block in header2.xml;
    // older ones use header1.xml. Try both, but only the present file
    // matters — `writeTemplateHeader` reads then writes the XML so the file
    // must exist. We swallow ENOENT for the missing variant.
    for (const headerName of ['header2.xml', 'header1.xml']) {
      const headerPath = join(unpackDir, 'word', headerName);
      try {
        await writeTemplateHeader(headerPath, {
          name: escaped.name,
          title_line1: escaped.title_line1,
          title_line2: escaped.title_line2,
          years: escaped.years,
        }, params.contract);
      } catch (err) {
        if ((err as NodeJS.ErrnoException).code !== 'ENOENT') {
          throw err;
        }
      }
    }

    const outDocxPath = join(workDir, 'out.docx');
    const [packOk, packMsg] = await packDocx(unpackDir, outDocxPath, baseDocxPath);
    if (!packOk) {
      throw new Error(`packDocx failed: ${packMsg}`);
    }

    const buf = await readFile(outDocxPath);
    return new Uint8Array(buf);
  });
}

/**
 * Public Contract 1 entry point. See module docstring for the pipeline.
 */
export async function transpose(input: TransposeInput): Promise<TransposeOutput> {
  const contract = manifestToContract(input.template.manifest, input.template.brand);
  const streamCallbacks = input.streamCallbacks;

  const results: TransposedCv[] = [];

  for (const file of input.files) {
    const errors: string[] = [];
    let warnings: string[] = [];
    let outputDocx: Uint8Array = new Uint8Array();
    let detectedFields: DetectedFields = {
      experienceCount: 0,
      educationCount: 0,
      skillBuckets: 0,
      languagesCount: 0,
    };
    let cvName = 'Candidate';
    let sourceText = '';
    let profile: CvData | null = null;
    let usage = { inputTokens: 0, outputTokens: 0, totalTokens: 0 };

    try {
      // 1. Extract raw text from the upload
      streamCallbacks?.onPhaseChange?.(file.name, 'extract-text');
      const rawText = await extractTextFromBuffer(Buffer.from(file.bytes), file.name);
      sourceText = rawText;

      // 2. LLM → CvData
      streamCallbacks?.onPhaseChange?.(file.name, 'extract-cv-llm');
      const llmResp = await input.llm.complete({
        systemPrompt: buildSystemPrompt(),
        userPrompt: buildUserPrompt({
          cvText: rawText,
          sourceFileName: file.name,
          userPromptOverride: file.userPromptOverride,
        }),
        maxTokens: 8192,
        temperature: 0.1,
        enableReasoning: input.extraction?.enableReasoning ?? true,
        reasoningBudget: input.extraction?.reasoningBudget,
        onDelta: streamCallbacks ? makeDeltaForwarder(file.name, streamCallbacks) : undefined,
      });
      if (llmResp.usage) {
        usage = {
          inputTokens: llmResp.usage.inputTokens,
          outputTokens: llmResp.usage.outputTokens,
          totalTokens: llmResp.usage.inputTokens + llmResp.usage.outputTokens,
        };
      }

      let parsed: unknown;
      try {
        parsed = JSON.parse(llmResp.text);
      } catch (err) {
        throw new Error(`LLM did not return valid JSON: ${(err as Error).message}`);
      }

      const cvDataResult = cvDataSchema.safeParse(parsed);
      if (!cvDataResult.success) {
        throw new Error(
          `LLM output failed schema validation: ${cvDataResult.error.issues.map((i) => `${i.path.join('.')}: ${i.message}`).join('; ')}`,
        );
      }
      profile = cvDataResult.data;
      cvName = profile.name;
      streamCallbacks?.onParsedKeys?.(file.name, Object.keys(profile));

      detectedFields = {
        name: profile.name,
        titleLine1: profile.title_line1,
        titleLine2: profile.title_line2,
        yearsOfExperience: profile.years ? Number.parseInt(profile.years, 10) || undefined : undefined,
        experienceCount: profile.experience.length,
        educationCount: profile.education.length,
        skillBuckets: profile.technicalSkills.length,
        languagesCount: profile.languages.length,
      };

      // 3. Render DOCX
      streamCallbacks?.onPhaseChange?.(file.name, 'render-docx');
      outputDocx = await renderDocxFromContract({
        baseDocx: input.template.baseDocx,
        data: profile,
        contract,
      });

      // 4. Validate page 1 (LibreOffice-based; produces warnings, not errors)
      streamCallbacks?.onPhaseChange?.(file.name, 'validate-page1');
      const v = await validatePage1(outputDocx, {
        experienceSectionLabel: getPrimaryExperienceSectionLabel(contract),
        sectorSectionLabel: getPrimarySectorSectionLabel(contract),
      });
      warnings = v.warnings;

      streamCallbacks?.onPhaseChange?.(file.name, 'done');
    } catch (err) {
      logger.error('transpose() failed for input', { file: file.name, err: (err as Error).message });
      errors.push((err as Error).message);
    }

    const outputDocxName = errors.length === 0
      ? deriveOutputNameFromTemplateContract(contract, file.name, cvName)
      : '';

    const alignmentReport: AlignmentReport = {
      validationPassed: errors.length === 0 && warnings.length === 0,
      warnings,
      detectedFields,
      page1SectionsFound: [],
      missingRequiredSections: [],
      retriesUsed: 0,
    };

    const safeProfile: CvData = profile ?? {
      name: cvName,
      title_line1: '',
      title_line2: '',
      years: '',
      technicalSkills: [],
      sectors: [],
      domains: [],
      experience: [],
      languages: [],
      education: [],
      attention_cv: '',
    };

    results.push({
      sourceFileName: file.name,
      outputDocxName,
      outputDocx,
      profile: safeProfile,
      sourceText,
      usage,
      alignmentReport,
      errors,
    });
  }

  return { results };
}

File: core/python/cv_transpose_core/transpose.py
from __future__ import annotations

from copy import deepcopy
import json
from typing import Any

from .contract import derive_output_name_from_contract, manifest_to_contract
from .docx import validate_base_docx
from .extract import extract_text_from_bytes
from .profile import EMPTY_PROFILE_FALLBACK, validate_cv_data
from .prompts import build_system_prompt, build_user_prompt
from .render import render_docx
from .types import (
    AlignmentReport,
    DetectedFields,
    LlmCompleteResult,
    TransposeInput,
    TransposedCv,
    TransposeOutput,
    Usage,
)
from .validate import validate_docx_structure, validate_page1

DEFAULT_EXTRACTION_MAX_TOKENS = 16_000
LARGE_CV_EXTRACTION_MAX_TOKENS = 32_000
LARGE_CV_TEXT_CHAR_THRESHOLD = 30_000
DEFAULT_PARSE_RETRIES = 1


def _fallback_profile() -> dict[str, Any]:
    return deepcopy(EMPTY_PROFILE_FALLBACK)


def _emit_phase(input_: TransposeInput, file_name: str, phase: str) -> None:
    if input_.stream_callbacks and input_.stream_callbacks.on_phase_change:
        input_.stream_callbacks.on_phase_change(file_name, phase)


def _on_delta(input_: TransposeInput, file_name: str):
    if input_.stream_callbacks is None:
        return None

    def route_delta(delta: dict[str, str]) -> None:
        kind = delta.get("kind")
        text = delta.get("text", "")
        if kind == "thinking" and input_.stream_callbacks and input_.stream_callbacks.on_thinking_delta:
            input_.stream_callbacks.on_thinking_delta(file_name, text)
        if kind == "content" and input_.stream_callbacks and input_.stream_callbacks.on_content_delta:
            input_.stream_callbacks.on_content_delta(file_name, text)

    return route_delta


def _empty_report() -> AlignmentReport:
    return AlignmentReport(
        validation_passed=False,
        warnings=[],
        detected_fields=DetectedFields(experience_count=0, education_count=0, skill_buckets=0, languages_count=0),
        page1_sections_found=[],
        missing_required_sections=[],
        retries_used=0,
    )


def _detected_fields(profile: dict[str, Any]) -> DetectedFields:
    years = None
    try:
        years = int(str(profile.get("years") or ""))
    except ValueError:
        years = None
    return DetectedFields(
        name=profile.get("name"),
        title_line1=profile.get("title_line1"),
        title_line2=profile.get("title_line2"),
        years_of_experience=years,
        experience_count=len(profile.get("experience", [])),
        education_count=len(profile.get("education", [])),
        skill_buckets=len(profile.get("technicalSkills", [])),
        languages_count=len(profile.get("languages", [])),
    )


def _primary_experience_section_label(contract: dict[str, Any]) -> str | None:
    for section in contract["sections"]:
        if section["key"] in {"experience", "selectedExperience", "additionalExperience"}:
            return section["label"]
    return None


def _primary_sector_section_label(contract: dict[str, Any]) -> str | None:
    for section in contract["sections"]:
        if section["key"] in {"sectorSkills", "sectorExperience"}:
            return section["label"]
    return None


def _coerce_llm_result(raw: LlmCompleteResult | dict[str, Any]) -> LlmCompleteResult:
    if isinstance(raw, LlmCompleteResult):
        return raw
    return LlmCompleteResult(text=raw["text"], usage=raw.get("usage"))


def _parse_llm_json(text: str) -> Any:
    trimmed = text.strip()
    without_fence = trimmed.removeprefix("```json").removeprefix("```").removesuffix("```").strip()
    return json.loads(without_fence)


def _parse_cv_data_from_llm(text: str) -> dict[str, Any]:
    try:
        parsed = _parse_llm_json(text)
    except json.JSONDecodeError as exc:
        raise ValueError(f"LLM did not return valid JSON: {exc.msg}: line {exc.lineno} column {exc.colno} (char {exc.pos})") from exc

    return validate_cv_data(parsed)


def _resolve_extraction_max_tokens(source_text: str, configured: int | None) -> int:
    if isinstance(configured, int) and configured > 0:
        return configured
    if len(source_text) > LARGE_CV_TEXT_CHAR_THRESHOLD:
        return LARGE_CV_EXTRACTION_MAX_TOKENS
    return DEFAULT_EXTRACTION_MAX_TOKENS


def _build_parse_retry_prompt(base_prompt: str | None, error: str) -> str:
    return "\n\n".join(
        part
        for part in [
            base_prompt or "",
            f"PREVIOUS LLM OUTPUT WAS NOT VALID JSON: {error}.",
            "Return ONLY one complete valid JSON object matching the requested schema. Do not include markdown, comments, analysis, or trailing text.",
        ]
        if part
    )


async def _process_one(file, input_: TransposeInput, contract: dict[str, Any]) -> TransposedCv:
    errors: list[str] = []
    source_text = ""
    profile: dict[str, Any] | None = None
    output_docx = b""
    report = _empty_report()
    usage_in = 0
    usage_out = 0
    retries_used = 0
    max_retries = input_.extraction.max_validation_retries if input_.extraction else None
    if max_retries is None:
        max_retries = 1
    max_parse_retries = input_.extraction.max_parse_retries if input_.extraction else None
    if max_parse_retries is None:
        max_parse_retries = DEFAULT_PARSE_RETRIES
    user_prompt_override = file.user_prompt_override

    try:
        validate_base_docx(input_.template.base_docx)
        _emit_phase(input_, file.name, "extract-text")
        source_text = extract_text_from_bytes(file.bytes_, file.name, file.mime)
        extraction_max_tokens = _resolve_extraction_max_tokens(
            source_text,
            input_.extraction.max_tokens if input_.extraction else None,
        )
        for attempt in range(max_retries + 1):
            if attempt > 0:
                _emit_phase(input_, file.name, "retry")
            parse_prompt_override = user_prompt_override
            profile = None
            for parse_attempt in range(max_parse_retries + 1):
                if parse_attempt > 0:
                    _emit_phase(input_, file.name, "retry")
                _emit_phase(input_, file.name, "extract-cv-llm")
                llm_result = _coerce_llm_result(
                    await input_.llm.complete(
                        system_prompt=build_system_prompt(),
                        user_prompt=build_user_prompt(
                            cv_text=source_text,
                            source_file_name=file.name,
                            user_prompt_override=parse_prompt_override,
                        ),
                        max_tokens=extraction_max_tokens,
                        temperature=0.1,
                        enable_reasoning=True if input_.extraction is None or input_.extraction.enable_reasoning is None else input_.extraction.enable_reasoning,
                        reasoning_budget=input_.extraction.reasoning_budget if input_.extraction else None,
                        response_format="json",
                        on_delta=_on_delta(input_, file.name),
                    )
                )
                if llm_result.usage:
                    usage_in += int(llm_result.usage.get("inputTokens", llm_result.usage.get("input_tokens", 0)))
                    usage_out += int(llm_result.usage.get("outputTokens", llm_result.usage.get("output_tokens", 0)))
                try:
                    profile = _parse_cv_data_from_llm(llm_result.text)
                    break
                except ValueError as exc:
                    if parse_attempt >= max_parse_retries:
                        raise
                    parse_prompt_override = _build_parse_retry_prompt(user_prompt_override, str(exc))
            if profile is None:
                raise ValueError("LLM extraction failed without a parsed profile")
            if input_.stream_callbacks and input_.stream_callbacks.on_parsed_keys:
                input_.stream_callbacks.on_parsed_keys(file.name, list(profile.keys()))
            _emit_phase(input_, file.name, "render-docx")
            output_docx = render_docx(
                input_.template.base_docx,
                profile,
                contract,
                renderer=input_.template.renderer or "generic",
            )
            required = [section["label"] for section in contract["sections"] if section["required"]]
            _emit_phase(input_, file.name, "validate-page1")
            page1 = validate_page1(
                output_docx,
                experience_section_label=_primary_experience_section_label(contract),
                sector_section_label=_primary_sector_section_label(contract),
            )
            structure = validate_docx_structure(output_docx, required)
            report = AlignmentReport(
                validation_passed=len(page1["warnings"]) == 0 and len(structure["missing"]) == 0,
                warnings=page1["warnings"],
                detected_fields=_detected_fields(profile),
                page1_sections_found=structure["found"],
                missing_required_sections=structure["missing"],
                retries_used=retries_used,
            )
            if report.validation_passed or attempt >= max_retries:
                break
            retries_used += 1
            feedback = "; ".join(
                [
                    *report.warnings,
                    *[f'Missing required section "{s}"' for s in report.missing_required_sections],
                ]
            )
            user_prompt_override = "\n\n".join(
                part
                for part in [
                    file.user_prompt_override or "",
                    f"VALIDATION ERRORS: {feedback}.",
                    "Shorten all skill descriptions to max 100 characters. Reduce sectors to max 4. Reduce domains to max 4.",
                ]
                if part
            )
        report.retries_used = retries_used
        _emit_phase(input_, file.name, "done")
    except Exception as exc:
        errors.append(str(exc))
        report.retries_used = retries_used

    safe_profile = profile or _fallback_profile()
    output_name = derive_output_name_from_contract(contract, file.name, safe_profile["name"]) if not errors else ""
    return TransposedCv(
        source_file_name=file.name,
        output_docx_name=output_name,
        output_docx=output_docx,
        profile=safe_profile,
        source_text=source_text,
        usage=Usage(input_tokens=usage_in, output_tokens=usage_out, total_tokens=usage_in + usage_out),
        alignment_report=report,
        errors=errors,
    )


async def transpose(input_: TransposeInput) -> TransposeOutput:
    contract = manifest_to_contract(input_.template.manifest, input_.template.brand)
    results = []
    for file in input_.files:
        results.append(await _process_one(file, input_, contract))
    return TransposeOutput(results=results)
